docs/requirements/ADP-957.md

```diff
@@ -0,0 +1,156 @@
+
+## Excessive collateral
+
+The distribution of a wallet's UTxO set may be such that all the pure-ada
+entries are far in excess of what would be required for collateral.
+
+In this case, the user might be reluctant to use any of their existing entries
+as collateral.
+
+## Potential solutions for these issues
+
+1.  Do nothing in addition to satisfying the requirements above.
+
+    If the user encounters a situation where their wallet's UTxO set does not
+    have enough collateral, or where the UTxO set only has pure-ada entries
+    that are too large, then they themselves will need to manually create a
+    rebalancing transaction that creates the appropriate collateral entries.
+
+2.  Adjust the wallet so that it tries to maintain, over time, a UTxO set that
+    has some proportion of entries suitable for inclusion as collateral inputs
+    in future transactions. This might involve the following adjustments:
+
+    a.  Adjust the migration algorithm so that it attempts to create a target
+        UTxO distribution with at least some entries suitable for use as
+        collateral inputs. This would probably require a relatively large
+        amount of development and testing effort to achieve, as the migration
+        algorithm is currently designed to do the opposite of this.
+
+    b.  Adjust the way we call the standard coin selection algorithm, so that
+        if the requested transaction would evolve the UTxO set to a state where
+        there are no entries suitable for use as collateral inputs, we
+        automatically append one or more additional pure-ada outputs (similar
+        to change outputs) to the transaction, to rectify this. This would
+        almost certainly add extra complexity to the coin selection algorithm.
+
+3.  On accepting a user request to make a selection that requires collateral:
+
+    a.  If the wallet detects that the UTxO set has **sufficient** collateral,
+        then return just one selection, with the outputs the user asked for.
+
+    b.  If the wallet detects that the UTxO set has **insufficient** collateral,
+        then return an ordered list with two selections:
+
+        1. Selection 1: a transaction that creates appropriate collateral.
+        2. Selection 2: a transaction corresponding to the user request.
+
+    c.  If the wallet detects that the UTxO set has **sufficient** collateral,
+        but that it's not possible to satisfy the `collateralPercentMaximum`
+        parameter, then return an ordered list with two selections:
+
+        1. Selection 1: a transaction that creates appropriate collateral.
+        2. Selection 2: a transaction corresponding to the user request.
+
+    In each of the above cases, the user can inspect the list of selections
+    returned, and decide whether or not they want to go ahead (by signing the
+    selections, and/or submitting them).
+
+4.  Provide an explicit endpoint that creates a collateral rearrangement
```

Comment:
I think that is relatively simple(st) solution which also provides pretty good user experience. It is simple as in: _when your tx fails you are advised to trigger this collateral ep first_ (which is far better than manually creating tx by themselves).

Reply:
@piotr-iohk wrote:
> I think that is relatively simple(st) solution which also provides pretty good user experience. It is simple as in: when your tx fails you are advised to trigger this collateral ep first (which is far better than manually creating tx by themselves).

I completely agree with you: it would be nice to provide the user with a convenient way to unblock themselves. Either via the explicit endpoint proposal (4), or the just-in-time rebalancing transaction proposal (3). 

Though I believe the priority at this stage is to deliver an MVP.

If we adopt proposal 1 (do nothing in addition to satisfying the basic requirements), then we can release something sooner that will at least allow advanced users to start working with collateral.